src/core/src/paged_cache_manager.cpp

```diff
@@ -0,0 +1,420 @@
+        if (pit != state.blocks.end()) {
+            const auto pos = static_cast<std::size_t>(std::distance(state.blocks.begin(), pit));
+            if (pos < state.scores.size())
+                state.scores[pos] = scores[i];
+        }
+        m_blocks[block_id].score = scores[i];
+    }
+    rebuild_evict_heap_unlocked();
+}
+
+// eviction
+void ov::util::PagedCacheManager::ensure_free_blocks_unlocked(std::size_t need_blocks) {
+    if (m_free_block_list.size() >= need_blocks)
+        return;
+    const std::size_t deficit = need_blocks - m_free_block_list.size();
+    for (std::size_t i = 0; i < deficit; ++i) {
+        evict_one_unlocked();
+        if (m_free_block_list.size() >= need_blocks)
+            break;
+    }
+}
+
+void ov::util::PagedCacheManager::evict_one_unlocked() {
+    if (m_evict_heap.empty())
+        rebuild_evict_heap_unlocked();
+    if (m_evict_heap.empty())
+        return;
+
+    std::pop_heap(m_evict_heap.begin(), m_evict_heap.end(), [this](std::size_t a, std::size_t b) {
+        return heap_less(m_blocks, a, b);
+    });
+    const std::size_t victim = m_evict_heap.back();
+    m_evict_heap.pop_back();
+
+    if (victim >= m_blocks.size())
+        return;
+    auto& pg = m_blocks[victim];
+    if (pg.owner == 0)
+        return;
+
+    const size_t owner = pg.owner;
+    auto it = m_ops.find(owner);
+    if (it != m_ops.end()) {
+        auto& state = it->second;
+        auto pit = std::find(state.blocks.begin(), state.blocks.end(), victim);
+        if (pit != state.blocks.end()) {
+            const auto pos = static_cast<std::size_t>(std::distance(state.blocks.begin(), pit));
+            state.blocks.erase(pit);
+            if (pos < state.scores.size())
+                state.scores.erase(state.scores.begin() + static_cast<std::ptrdiff_t>(pos));
+        }
+    }
+
+    pg.owner = 0;
+    pg.score = std::numeric_limits<float>::infinity();
+    m_free_block_list.push_back(victim);
+}
+
+void ov::util::PagedCacheManager::evict_to_target_free(std::size_t target_free_blocks) {
+    std::lock_guard<std::mutex> lock(m_mutex);
+    while (m_free_block_list.size() < target_free_blocks) {
+        evict_one_unlocked();
+        if (m_evict_heap.empty())
+            break;
+    }
+}
+
+// priv helpers
+void* ov::util::PagedCacheManager::offset_key(std::size_t block_idx) const noexcept {
+    return static_cast<void*>(static_cast<unsigned char*>(m_key_base) + block_idx * m_block_bytes);
+}
+
+void* ov::util::PagedCacheManager::offset_value(std::size_t block_idx) const noexcept {
+    return static_cast<void*>(static_cast<unsigned char*>(m_value_base) + block_idx * m_block_bytes);
+}
+
+void ov::util::PagedCacheManager::compute_subsequence_begins_unlocked(operator_state& state) const {
+    (void)state;  // hook for real PA wiring if needed
+}
+
+float ov::util::PagedCacheManager::cast_score_to_float(ov::element::Type et, const void* src_scalar) noexcept {
```

Comment:
OV operators not should provide such conversion/cast
Check function in f16 bf16 types and convert implementation

Reply:
Fixed!